pkg/controller/egress/controller.go

```diff
@@ -274,29 +278,29 @@ func (c *EgressController) syncEgressIP(egress *egressv1beta1.Egress) (net.IP, e
 	if egress.Spec.EgressIP != "" {
 		ip = net.ParseIP(egress.Spec.EgressIP)
 		if err := c.externalIPAllocator.UpdateIPAllocation(egress.Spec.ExternalIPPool, ip); err != nil {
-			return nil, fmt.Errorf("error when allocating IP %v for Egress %s from ExternalIPPool %s: %v", ip, egress.Name, egress.Spec.ExternalIPPool, err)
+			return nil, updatedEgress, fmt.Errorf("error when allocating IP %v for Egress %s from ExternalIPPool %s: %v", ip, egress.Name, egress.Spec.ExternalIPPool, err)
 		}
 	} else {
 		var err error
 		// User doesn't specify the Egress IP, allocate one.
 		if ip, err = c.externalIPAllocator.AllocateIPFromPool(egress.Spec.ExternalIPPool); err != nil {
-			return nil, err
+			return nil, updatedEgress, err
 		}
-		if err = c.updateEgressIP(egress, ip.String()); err != nil {
+		if updatedEgress, err = c.updateEgressIP(egress, ip.String()); err != nil {
 			if rerr := c.externalIPAllocator.ReleaseIP(egress.Spec.ExternalIPPool, ip); rerr != nil &&
 				rerr != externalippool.ErrExternalIPPoolNotFound {
 				klog.ErrorS(rerr, "Failed to release IP", "ip", ip, "pool", egress.Spec.ExternalIPPool)
 			}
-			return nil, err
+			return nil, updatedEgress, err
```

Comment:
Note that `updateEgressIP` could be `nil`, and the program would panic when `updateEgressAllocatedCondition` accesses `egress.Spec.ExternalIPPool`.
I feel you could declare `updatedEgress` only when necessary. For example:
```
func (c *EgressController) syncEgressIP(egress *egressv1beta1.Egress) (net.IP, *egressv1beta1.Egress, error) {
	prevIP, prevIPPool, exists := c.getIPAllocation(egress.Name)
	if exists {
		// The EgressIP and the ExternalIPPool don't change, do nothing.
		if prevIP.String() == egress.Spec.EgressIP && prevIPPool == egress.Spec.ExternalIPPool && c.externalIPAllocator.IPPoolExists(egress.Spec.ExternalIPPool) {
			return prevIP, egress, nil
		}
		// Either EgressIP or ExternalIPPool changes, release the previous one first.
		if err := c.releaseEgressIP(egress.Name, prevIP, prevIPPool); err != nil {
			return nil, egress, err
		}
	}
	...
	if !c.externalIPAllocator.IPPoolExists(egress.Spec.ExternalIPPool) {
		// The IP pool has been deleted, reclaim the IP from the Egress API.
		if egress.Spec.EgressIP != "" {
			if updatedEgress, err := c.updateEgressIP(egress, ""); err != nil {
				return nil, egress, err
			} else {
				egress = updatedEgress
			}
		}
		return nil, egress, fmt.Errorf("ExternalIPPool %s does not exist", egress.Spec.ExternalIPPool)
	}
	...
```
In this way, we only override egress when it's really updated.